---
title: "assignment"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Part I

## Data Inspection

```{r}
getwd()
setwd("/Users/huyusmac/Documents/class/BCB546X/546class/R_assignment")
file.info("fang_et_al_genotypes.txt")
file.info("snp_position.txt")
genotype <- read.csv("fang_et_al_genotypes.txt",header = FALSE, sep = '\t')
snp <- read.csv("snp_position.txt",header = TRUE, sep = '\t')
head(genotype,10)
tail(genotype,10)
str(genotype)
colnames(genotype)
rownames(genotype)
dim(genotype)
head(snp,10)
tail(snp,10)
str(snp)
colnames(snp)
rownames(snp)
dim(snp)
#summary(genotype)
#summary(snp)
```

By inspecting these two file I learned that:
* `fang_et_al_genotypes.txt` has 11051939 bytes, 986 column, 2783 rows, and the first row is header.The snp data is start from the forth column.
* `snp_position.txt` has 82763 bytes, 15 column, 984 rows, and the first row is header.The snp data is start from the second row.

## Data Processing

```{r}
library(dplyr)
genotype <- as.data.frame(genotype)
snp <- as.data.frame(snp)
extra <- data_frame(c("JG_OTU",rep("_",14)),c("Group",rep("_",14)))
extra <- as.data.frame(t(extra))
rownames(extra) <- c()
colnames(extra) <- colnames(snp)
snp <- rbind(snp,extra)
snp <- snp[order(snp$SNP_ID),]
snp <- as.data.frame(snp)
```

I added extra two rows to snp file, and ordered it.

###Maize Data

```{r}
maize <- subset(genotype, genotype$V3 %in% c("Group","ZMMLR", "ZMMMR", "ZMMIL"))
maize <- t(maize)
maize[1,1] <- "SNP_ID"
colnames(maize) <- maize[1,]
maize <- as.data.frame(maize)
maize <- maize[-1,]
maize <- maize[order(maize$SNP_ID),]
maize <- as.data.frame(maize)
maize_all <- merge(snp, maize, by = "SNP_ID")
maize_all <- maize_all[c(14,16,1:13,15,17:985),]

multi_maize <- subset(maize_all, maize_all$mult_positions != "")
multi_maize <- subset(multi_maize, multi_maize$mult_positions != "_")
write.table(multi_maize, file = "maize_multi.txt", sep = "\t", row.names = FALSE)
unknown_maize <- subset(maize_all, maize_all$Position == "unknown")
write.table(unknown_maize, file = "maize_unknown.txt", sep = "\t", row.names = FALSE)

single_maize<-maize_all[-which(maize_all$mult_positions != ""),]
single_maize <- maize_all[which(single_maize$Chromosome != "unknown"),]
splitData<-split(single_maize, single_maize$Chromosome)
allNames <- names(splitData)
for(thisName in allNames[c(1:10)]){
     splitData[[thisName]] <- splitData[[thisName]][order(as.numeric(as.character(splitData[[thisName]]$Position))),]
     saveName = paste0("maize_", thisName, '.txt')
     write.table(splitData[[thisName]], file = saveName, sep = "\t", row.names = FALSE)
}
library(tidyverse)
single_maize_r <- data.frame(lapply(single_maize, as.character), stringsAsFactors=FALSE)
single_maize_r <- data.frame(sapply(single_maize_r,function(x) {x <- gsub("?","-",x,fixed=TRUE)}))
for(thisName in allNames[c(1:10)]){
     splitData[[thisName]] <- splitData[[thisName]][order(as.numeric(as.character(splitData[[thisName]]$Position)), decreasing = TRUE),]
     saveName = paste0("maize_r_", thisName, '.txt')
     write.table(splitData[[thisName]], file = saveName, sep = "\t", row.names = FALSE)
}
```

Maize data was selected from whole file, then combined with snp position data. Multiple position SNPs and unknown position SNPs were firstly separated. Then I separated rest SNPs into each file based on chromosome and sorted by position. I also generated files of each chromosome with decreasing position and missing data encoded by `-`, these files were named with `r` in the middle.

###Teosinte Data

```{r}
teosinte <- subset(genotype, genotype$V3 %in% c("Group","ZMPBA", "ZMPIL", "ZMPJA"))
teosinte <- t(teosinte)
teosinte[1,1] <- "SNP_ID"
colnames(teosinte) <- teosinte[1,]
teosinte <- as.data.frame(teosinte)
teosinte <- teosinte[-1,]
teosinte <- teosinte[order(teosinte$SNP_ID),]
teosinte <- as.data.frame(teosinte)
teosinte_all <- merge(snp, teosinte, by = "SNP_ID")
teosinte_all <- teosinte_all[c(14,16,1:13,15,17:985),]
multi_teosinte <- subset(teosinte_all, teosinte_all$mult_positions != "")
multi_teosinte <- subset(multi_teosinte, multi_teosinte$mult_positions != "_")
write.table(multi_teosinte, file = "teosinte_multi.txt", sep = "\t", row.names = FALSE)
unknown_teosinte <- subset(teosinte_all, teosinte_all$Position == "unknown")
write.table(unknown_teosinte, file = "teosinte_unknown.txt", sep = "\t", row.names = FALSE)

single_teosinte<-teosinte_all[-which(teosinte_all$mult_positions != ""),]
single_teosinte <- teosinte_all[which(single_teosinte$Chromosome != "unknown"),]
splitData<-split(single_teosinte, single_teosinte$Chromosome)
allNames <- names(splitData)
for(thisName in allNames[c(1:10)]){
     splitData[[thisName]] <- splitData[[thisName]][order(as.numeric(as.character(splitData[[thisName]]$Position))),]
     saveName = paste0("teosinte_", thisName, '.txt')
     write.table(splitData[[thisName]], file = saveName, sep = "\t", row.names = FALSE)
}
single_teosinte_r <- data.frame(lapply(single_teosinte, as.character), stringsAsFactors=FALSE)
single_teosinte_r <- data.frame(sapply(single_teosinte_r,function(x) {x <- gsub("?","-",x,fixed=TRUE)}))
for(thisName in allNames[c(1:10)]){
     splitData[[thisName]] <- splitData[[thisName]][order(as.numeric(as.character(splitData[[thisName]]$Position)), decreasing = TRUE),]
     saveName = paste0("teosinte_r_", thisName, '.txt')
     write.table(splitData[[thisName]], file = saveName, sep = "\t", row.names = FALSE)
}
```

Same with maize data, but work on teosinte data.


# Part II

```{r}
library(ggplot2)
```

## SNPs per chromosome

```{r}
snp <- snp[c(1:983),]
snp_summary <- as.data.frame(summary(snp$Chromosome))
snp_summary <- cbind(rownames(snp_summary),snp_summary)
colnames(snp_summary) <- c("chromosome","count")
snp_summary <- snp_summary[order(as.numeric(as.character(snp_summary$chromosome))),]
snp_summary <- snp_summary[c(1:12),]
ggplot(data=snp_summary, aes(x=chromosome, y=count)) + geom_point() + scale_x_discrete(limits=c(1:10,"multiple","unknown"))
ggplot(data=snp, aes(x=Chromosome, y=order(as.numeric(as.character(Position))))) + geom_point() + scale_x_discrete(limits=c(1:10)) + ylab("Position")
```

The total number of SNPs on each chromosome was counted and ploted.
The distribution of SNPs on chromosomes was ploted.

## Missing data and amount of heterozygosity

```{r}
genotype2 <- data.frame(lapply(genotype, as.character), stringsAsFactors=FALSE)
genotype2 <- data.frame(sapply(genotype2,function(x) {x <- gsub("([ACGT])/\\1","homo",x)}))
genotype2 <- data.frame(sapply(genotype2,function(x) {x <- gsub("?/?","missing",x,fixed=TRUE)}))
genotype2 <- data.frame(sapply(genotype2,function(x) {x <- gsub("\\w/\\w","hete",x)}))
genotype3 <- genotype2[-1,]
genotype3 <- genotype3[,c(4:986)]
genotype3 <- data.frame(lapply(genotype3, as.character), stringsAsFactors=FALSE)
snp_sum <- as.data.frame(sapply(genotype3, function(x) {table(x)}))
snp_sum <- as.data.frame(t(snp_sum))
snp_sum <- cbind(t(genotype2[1,c(4:986)]), snp_sum)
genotype4 <- as.data.frame(t(genotype3))
genotype4 <- data.frame(lapply(genotype4, as.character), stringsAsFactors=FALSE)
sample_sum <- sapply(genotype4,function(x) {table(x)})
sample_sum <- data.frame(matrix(unlist(sample_sum), nrow=length(sample_sum), byrow=T))
sample_sum<-cbind(genotype2[-1,c(1,3)], sample_sum)
colnames(sample_sum)<-c("Sample","Group","hete","homo","missing")
sample1<-sample_sum[,c(1:3)]
sample1 <- cbind(sample1,"hete")
colnames(sample1)<-c("Sample","Group","count","genotype")
sample2<-sample_sum[,c(1,2,4)]
sample2 <- cbind(sample2,"homo")
colnames(sample2)<-c("Sample","Group","count","genotype")
sample3<-sample_sum[,c(1,2,5)]
sample3 <- cbind(sample3,"missing")
colnames(sample3)<-c("Sample","Group","count","genotype")
sample_sum<- rbind(sample1,sample2,sample3)
ggplot(data=sample_sum, aes(x=Sample, y=count, fill=genotype)) + geom_bar(stat="identity", position = 'fill') + theme(axis.text.x = element_blank(), axis.ticks.x=element_blank()) +ylab("proportion")
ggplot(data=sample_sum, aes(x=Group, y=count, fill=genotype)) + geom_bar(stat="identity", position = 'fill') + theme(axis.text.x = element_text(vjust = 0.5, hjust = 0.5, angle = 45))+ylab("proportion")
```

The data was inverted by only showing homozygous `homo`, heterozygous `hete` and missing data `missing` instead of particular nucleotide.
Then it is summaried based on SNPs and samples separately.
The data of summary of each sample was later used to plot proportion of homozygous and heterozygous sites as well as missing data in each sample and group.

## Maize vs teosinte

```{r}
maize_sum <- subset(sample_sum, sample_sum$Group %in% c("ZMMLR", "ZMMMR", "ZMMIL"))
teosinte_sum <- subset(sample_sum, sample_sum$Group %in% c("ZMPBA", "ZMPIL", "ZMPJA"))
library(plyr)
maize_sum<-ddply(maize_sum,.(genotype),summarise,sum_count=sum(count))
maize_sum$Group<-"maize"
teosinte_sum<-ddply(teosinte_sum,.(genotype),summarise,sum_count=sum(count))
teosinte_sum$Group<-"teosinte"
maize_vs_teosinte<-rbind(maize_sum,teosinte_sum)
ggplot(data=maize_vs_teosinte, mapping=aes(x=Group, y=sum_count, fill=genotype)) + geom_bar(stat="identity", position = 'stack')+ geom_text(mapping = aes(label = sum_count),size = 5, colour = 'black',  position = position_stack())
```

I also compared the number of homozygous and heterozygous sites as well as missing data in maize and teosinte.
